Voir un message où on est pingé donne une 500

```
[ERROR] -- 2017-02-11 14:11:25,507 -- base : Internal Server Error: /forums/sujet/7920/salidou/
Traceback (most recent call last):
  File "/opt/zds/zdsenv/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/zds/zdsenv/local/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/zds/zdsenv/local/lib/python2.7/site-packages/django/views/generic/base.py", line 89, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/zds/zds-site/zds/forum/views.py", line 152, in get
    return super(TopicPostsListView, self).get(request, *args, **kwargs)
  File "/opt/zds/zdsenv/local/lib/python2.7/site-packages/django/views/generic/list.py", line 174, in get
    context = self.get_context_data()
  File "/opt/zds/zds-site/zds/forum/views.py", line 186, in get_context_data
    signals.content_read.send(sender=post.__class__, instance=post, user=self.request.user)
  File "/opt/zds/zdsenv/local/lib/python2.7/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/opt/zds/zds-site/zds/notification/receivers.py", line 145, in mark_comment_read
    subscription.mark_notification_read(comment)
  File "/opt/zds/zds-site/zds/notification/models.py", line 226, in mark_notification_read
    LOG.warning('%s notifications were find for %s/%s', len(notifications), content.type, content.title)
AttributeError: 'Post' object has no attribute 'type'
```

Probablement facile à réparer.

--- a/zds/notification/managers.py
+++ b/zds/notification/managers.py
@@ -156,6 +156,16 @@ class SubscriptionManager(models.Manager):
             existing.deactivate()
         return existing
 
+    def deactivate_subscriptions(self, user, _object):
+        subscription = self.get_existing(user, _object)
+        if subscription:
+            subscription.is_active = False
+            notification = subscription.last_notification
+            notification.is_read = True
+            notification.is_dead = True
+            notification.save(update_fields=['is_read', 'is_dead'])
+            subscription.save(update_fields=['is_active'])
+
 
 class NewTopicSubscriptionManager(SubscriptionManager):
     def mark_read_everybody_at(self, topic):

--- a/zds/notification/models.py
+++ b/zds/notification/models.py
@@ -10,7 +10,7 @@ from django.template.loader import render_to_string
 from django.utils.translation import ugettext_lazy as _
 from django.conf import settings
 
-from zds.forum.models import Topic
+from zds.forum.models import Topic, Post
 from zds.notification.managers import NotificationManager, SubscriptionManager, TopicFollowedManager, \
     TopicAnswerSubscriptionManager, NewTopicSubscriptionManager
 from zds.utils.misc import convert_camel_to_underscore
@@ -332,6 +332,16 @@ class PingSubscription(AnswerSubscription, MultipleNotificationsMixin):
 
     def __str__(self):
         return _('<Abonnement du membre "{0}" aux mentions>').format(self.profile, self.object_id)
+
+    @staticmethod
+    def mark_inaccessible_ping_as_read_for_topic(topic):
+        for post in Post.objects.filter(topic=topic):
+            for notification in Notification.objects.filter(content_type__pk=ContentType.objects.get_for_model(post).pk,
+                                                            object_id=post.pk):
+                if not topic.forum.can_read(notification.subscription.user):
+                    notification.is_read = True
+                    notification.is_dead = True
+                    notification.save(update_fields=['is_read'])
 
     def get_notification_title(self, answer):
         assert hasattr(answer, 'author')

--- a/zds/notification/models.py
+++ b/zds/notification/models.py
@@ -220,7 +220,9 @@ class MultipleNotificationsMixin(object):
                     LOG.debug('nothing to mark as read')
                     return
                 elif len(notifications) > 1:
-                    LOG.warning('%s notifications found for %s/%s', len(notifications), content.type, content.title)
+                    content_type = getattr(content, 'type', 'forum')
+                    content_title = getattr(content, 'title', 'message')
+                    LOG.warning('%s notifications found for %s/%s', len(notifications), content_type, content_title)
                     for notif in notifications[1:]:
                         notif.delete()
 

--- a/zds/notification/receivers.py
+++ b/zds/notification/receivers.py
@@ -4,7 +4,6 @@ import logging
 
 from django.db import DatabaseError
 
-import zds
 from zds.tutorialv2.signals import content_unpublished
 
 from zds.utils.models import Tag
@@ -22,7 +21,7 @@ from zds.mp.models import PrivateTopic, PrivatePost
 from zds.notification.models import TopicAnswerSubscription, ContentReactionAnswerSubscription, \
     PrivateTopicAnswerSubscription, Subscription, Notification, NewTopicSubscription, NewPublicationSubscription, \
     PingSubscription
-from zds.notification.signals import answer_unread, content_read, new_content, edit_content
+from zds.notification.signals import answer_unread, content_read, new_content, edit_content, unsubscribe
 from zds.tutorialv2.models.database import PublishableContent, ContentReaction
 import zds.tutorialv2.signals
 
@@ -469,3 +468,9 @@ def cleanup_notification_for_unpublished_content(sender, instance, **__):
         logger.debug('Nothing went wrong.')
     except DatabaseError as e:
         logger.exception('Error while saving %s, %s', instance, e)
+
+
+@receiver(unsubscribe)
+def unsubscripte_unpinged_user(sender, instance, user, **_):
+    if user:
+        PingSubscription.objects.deactivate_subscriptions(user, instance)

--- a/zds/notification/receivers.py
+++ b/zds/notification/receivers.py
@@ -104,12 +104,12 @@ def mark_topic_notifications_read(sender, *, instance, user, **__):
             subscription.mark_notification_read(content=instance)
 
     content_type = ContentType.objects.get_for_model(instance)
-    notifications = Notification.objects.filter(subscription__user=user, object_id=instance.pk,
-                                                content_type__pk=content_type.pk, is_read=False,
-                                                is_dead=True)
+    notifications = list(Notification.objects.filter(subscription__user=user, object_id=instance.pk,
+                                                     content_type__pk=content_type.pk, is_read=False))
+
     for notification in notifications:
         notification.is_read = True
-        notification.save()
+        notification.save(update_fields=['is_read'])
 
 
 @receiver(content_read, sender=PublishableContent)
@@ -181,7 +181,7 @@ def edit_topic_event(sender, *, action, instance, **kwargs):
 
     if action == 'move':
 
-        _handle_private_forum_moving(topic, topic_content_type)
+        _handle_private_forum_moving(topic, topic_content_type, ContentType.objects.get_for_model(topic.last_message))
 
     elif action == 'edit_tags_and_title':
         topic = instance
@@ -226,24 +226,24 @@ def _handle_deleted_tags(topic, topic_content_type):
     return tag_content_type
 
 
-def _handle_private_forum_moving(topic, topic_content_type):
+def _handle_private_forum_moving(topic, topic_content_type, post_content_type):
     # If the topic is moved to a restricted forum, users who cannot read this topic any more unfollow it.
     # This avoids unreachable notifications.
     TopicAnswerSubscription.objects.unfollow_and_mark_read_everybody_at(topic)
     NewTopicSubscription.objects.mark_read_everybody_at(topic)
+    PingSubscription.mark_inaccessible_ping_as_read_for_topic(topic)
     # If the topic is moved to a forum followed by the user, we update the subscription of the notification.
     # Otherwise, we update the notification as dead.
-    notifications = Notification.objects \
-        .filter(object_id=topic.pk, content_type__pk=topic_content_type.pk, is_read=False).all()
+    notifications = list(Notification.objects
+                         .filter(object_id=topic.pk, content_type__pk=topic_content_type.pk, is_read=False).all())
     for notification in notifications:
-        subscription = NewTopicSubscription.objects \
-            .get_existing(notification.subscription.user, topic.forum, is_active=True)
-        if subscription:
+        subscription = notification.subscription
+        if subscription.is_active:
             notification.subscription = subscription
             notification.save()
         elif notification.subscription.content_object != notification.content_object.forum:
             notification.is_dead = True
-            notification.save()
+            notification.save(update_fields=['is_dead', 'is_read'])
 
 
 @receiver(post_save, sender=Topic)
@@ -347,7 +347,8 @@ def answer_comment_event(sender, *, instance, user, **__):
 
     assert comment is not None
     assert user is not None
-
+    if not instance.topic.forum.can_read(user):
+        return
     subscription = PingSubscription.objects.get_or_create_active(user, comment)
     subscription.send_notification(content=comment, sender=comment.author, send_email=False)
 

--- a/zds/notification/signals.py
+++ b/zds/notification/signals.py
@@ -12,3 +12,5 @@ edit_content = Signal(providing_args=['instance', 'action'])
 
 # is sent when a content is read (topic, article or tutorial)
 content_read = Signal(providing_args=['instance', 'user', 'target'])
+
+unsubscribe = Signal(providing_args=['instance', 'user'])

--- a/zds/notification/tests/tests_basics.py
+++ b/zds/notification/tests/tests_basics.py
@@ -334,14 +334,14 @@ class NotificationForumTest(TestCase):
 
         topic = Topic.objects.get(pk=topic.pk)
         self.assertEqual(self.forum12, topic.forum)
-        self.assertEqual(1, len(Notification.objects.filter(object_id=topic.pk, is_read=False, is_dead=True).all()))
+        self.assertEqual(1, len(Notification.objects.filter(object_id=topic.pk, is_read=False).all()))
 
         self.client.logout()
         self.assertTrue(self.client.login(username=self.user1.username, password='hostel77'))
         response = self.client.get(reverse('topic-posts-list', args=[topic.pk, topic.slug()]))
         self.assertEqual(200, response.status_code)
 
-        self.assertEqual(1, len(Notification.objects.filter(object_id=topic.pk, is_read=True, is_dead=True).all()))
+        self.assertEqual(1, len(Notification.objects.filter(object_id=topic.pk, is_read=True).all()))
 
     def test_move_topic_from_forum_followed_to_forum_followed_too(self):
         NewTopicSubscription.objects.toggle_follow(self.forum11, self.user1)

--- a/zds/notification/tests/tests_tricky.py
+++ b/zds/notification/tests/tests_tricky.py
@@ -6,7 +6,7 @@ from django.test import TestCase
 from django.test.utils import override_settings
 
 from zds.forum.factories import CategoryFactory, ForumFactory
-from zds.forum.models import Topic, Post
+from zds.forum.models import Topic
 from zds.gallery.factories import UserGalleryFactory
 from zds.member.factories import StaffProfileFactory, ProfileFactory
 from zds.notification.models import NewTopicSubscription, Notification, NewPublicationSubscription, \
@@ -65,7 +65,7 @@ class ForumNotification(TestCase):
             {
                 'title': 'Super sujet',
                 'subtitle': 'Pour tester les notifs',
-                'text': "@{} is pinged".format(self.user1.username),
+                'text': '@{} is pinged'.format(self.user1.username),
                 'tags': ''
             },
             follow=False)
@@ -131,6 +131,58 @@ class ForumNotification(TestCase):
         response = self.client.post(reverse('topic-edit'), data, follow=False)
         self.assertEqual(302, response.status_code)
         subscription = NewTopicSubscription.objects.get_existing(self.user1, self.forum11, True)
+        self.assertIsNotNone(subscription, 'There must still be an active subscription')
+        self.assertIsNotNone(subscription.last_notification,
+                             'There must still be a notification as object is not removed.')
+        self.assertEqual(subscription.last_notification,
+                         Notification.objects.filter(sender=self.user2).first())
+        self.assertTrue(subscription.last_notification.is_read, 'As forum is not reachable, notification is read')
+
+    def test_no_ping_on_private_forum(self):
+        self.assertTrue(self.client.login(username=self.staff.username, password='hostel77'))
+        result = self.client.post(
+            reverse('topic-new') + '?forum={0}'.format(self.forum12.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': 'ping @{}'.format(self.user1.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+
+        topic = Topic.objects.filter(title='Super sujet').first()
+        subscription = PingSubscription.objects.get_existing(self.user1, topic.last_message, True)
+        self.assertIsNone(subscription, 'There must be no active subscription for now')
+
+    def test_no_dead_ping_notif_on_moving_to_private_forum(self):
+        self.assertTrue(self.client.login(username=self.user2.username, password='hostel77'))
+        result = self.client.post(
+            reverse('topic-new') + '?forum={0}'.format(self.forum11.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': 'ping @{}'.format(self.user1.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+
+        topic = Topic.objects.filter(title='Super sujet').first()
+        subscription = PingSubscription.objects.get_existing(self.user1, topic.last_message, True)
+        self.assertIsNotNone(subscription, 'There must be an active subscription for now')
+        self.assertIsNotNone(subscription.last_notification, 'There must be a notification for now')
+        self.assertFalse(subscription.last_notification.is_read)
+        self.client.logout()
+        self.assertTrue(self.client.login(username=self.staff.username, password='hostel77'))
+        data = {
+            'move': '',
+            'forum': self.forum12.pk,
+            'topic': topic.pk
+        }
+        response = self.client.post(reverse('topic-edit'), data, follow=False)
+        self.assertEqual(302, response.status_code)
+        subscription = PingSubscription.objects.get_existing(self.user1, topic.last_message, True)
         self.assertIsNotNone(subscription, 'There must still be an active subscription')
         self.assertIsNotNone(subscription.last_notification,
                              'There must still be a notification as object is not removed.')

--- a/zds/notification/tests/tests_tricky.py
+++ b/zds/notification/tests/tests_tricky.py
@@ -6,7 +6,7 @@ from django.test import TestCase
 from django.test.utils import override_settings
 
 from zds.forum.factories import CategoryFactory, ForumFactory
-from zds.forum.models import Topic
+from zds.forum.models import Topic, Post
 from zds.gallery.factories import UserGalleryFactory
 from zds.member.factories import StaffProfileFactory, ProfileFactory
 from zds.notification.models import NewTopicSubscription, Notification, NewPublicationSubscription, \
@@ -46,13 +46,61 @@ class ForumNotification(TestCase):
             {
                 'title': 'Super sujet',
                 'subtitle': 'Pour tester les notifs',
-                'text': "@{} is pinged, not @{}".format(self.user1.username, self.user2.username),
+                'text': '@{} is pinged, not @{}'.format(self.user1.username, self.user2.username),
                 'tags': ''
             },
             follow=False)
         self.assertEqual(result.status_code, 302)
         self.assertEqual(1, PingSubscription.objects.count(),
                          'As one user is pinged, only one subscription is created.')
+
+    def test_no_reping_on_edition(self):
+        """
+        to be more accurate : on edition, only ping **new** members
+        """
+        overridden_zds_app['comment']['enable_pings'] = True
+        self.assertTrue(self.client.login(username=self.user2.username, password='hostel77'))
+        result = self.client.post(
+            reverse('topic-new') + '?forum={0}'.format(self.forum11.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': "@{} is pinged".format(self.user1.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(1, PingSubscription.objects.count(),
+                         'As one user is pinged, only one subscription is created.')
+        self.assertEqual(1, Notification.objects.count())
+        user3 = ProfileFactory().user
+        post = Topic.objects.last().last_message
+        result = self.client.post(
+            reverse('post-edit') + '?message={0}'.format(post.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': '@{} is pinged even twice @{}'.format(self.user1.username, self.user1.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(1, PingSubscription.objects.count(),
+                         'No added subscription.')
+        self.assertEqual(1, Notification.objects.count())
+        result = self.client.post(
+            reverse('post-edit') + '?message={0}'.format(post.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': '@{} is pinged even twice @{} and add @{}'.format(self.user1.username,
+                                                                          self.user1.username, user3.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(2, PingSubscription.objects.count())
+        self.assertEqual(2, Notification.objects.count())
 
     def test_no_dead_notif_on_moving(self):
         NewTopicSubscription.objects.get_or_create_active(self.user1, self.forum11)

--- a/zds/notification/tests/tests_tricky.py
+++ b/zds/notification/tests/tests_tricky.py
@@ -102,6 +102,42 @@ class ForumNotification(TestCase):
         self.assertEqual(2, PingSubscription.objects.count())
         self.assertEqual(2, Notification.objects.count())
 
+    def test_loose_ping_on_edition(self):
+        """
+        to be more accurate : on edition, only ping **new** members
+        """
+        overridden_zds_app['comment']['enable_pings'] = True
+        self.assertTrue(self.client.login(username=self.user2.username, password='hostel77'))
+        result = self.client.post(
+            reverse('topic-new') + '?forum={0}'.format(self.forum11.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': '@{} is pinged'.format(self.user1.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(1, PingSubscription.objects.count(),
+                         'As one user is pinged, only one subscription is created.')
+        self.assertEqual(1, Notification.objects.count())
+        post = Topic.objects.last().last_message
+        result = self.client.post(
+            reverse('post-edit') + '?message={0}'.format(post.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': '@ {} is no more pinged '.format(self.user1.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(1, PingSubscription.objects.count(),
+                         'No added subscription.')
+        self.assertFalse(PingSubscription.objects.first().is_active)
+        self.assertEqual(1, Notification.objects.count())
+        self.assertTrue(Notification.objects.first().is_read)
+
     def test_no_dead_notif_on_moving(self):
         NewTopicSubscription.objects.get_or_create_active(self.user1, self.forum11)
         self.assertTrue(self.client.login(username=self.user2.username, password='hostel77'))

--- a/zds/notification/tests/tests_tricky.py
+++ b/zds/notification/tests/tests_tricky.py
@@ -10,7 +10,7 @@ from zds.forum.models import Topic
 from zds.gallery.factories import UserGalleryFactory
 from zds.member.factories import StaffProfileFactory, ProfileFactory
 from zds.notification.models import NewTopicSubscription, Notification, NewPublicationSubscription, \
-    ContentReactionAnswerSubscription
+    ContentReactionAnswerSubscription, PingSubscription
 from zds.notification import signals as notif_signals
 from zds.tutorialv2.factories import PublishableContentFactory, LicenceFactory, SubCategoryFactory, \
     PublishedContentFactory, ContentReactionFactory
@@ -20,7 +20,10 @@ from copy import deepcopy
 
 from zds.utils.templatetags.interventions import interventions_topics
 
+overridden_zds_app = deepcopy(settings.ZDS_APP)
 
+
+@override_settings(ZDS_APP=overridden_zds_app)
 class ForumNotification(TestCase):
     def setUp(self):
         self.user1 = ProfileFactory().user
@@ -34,6 +37,22 @@ class ForumNotification(TestCase):
         for group in self.staff.groups.all():
             self.forum12.groups.add(group)
         self.forum12.save()
+
+    def test_no_auto_ping(self):
+        overridden_zds_app['comment']['enable_pings'] = True
+        self.assertTrue(self.client.login(username=self.user2.username, password='hostel77'))
+        result = self.client.post(
+            reverse('topic-new') + '?forum={0}'.format(self.forum11.pk),
+            {
+                'title': 'Super sujet',
+                'subtitle': 'Pour tester les notifs',
+                'text': "@{} is pinged, not @{}".format(self.user1.username, self.user2.username),
+                'tags': ''
+            },
+            follow=False)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(1, PingSubscription.objects.count(),
+                         'As one user is pinged, only one subscription is created.')
 
     def test_no_dead_notif_on_moving(self):
         NewTopicSubscription.objects.get_or_create_active(self.user1, self.forum11)

--- a/zds/utils/models.py
+++ b/zds/utils/models.py
@@ -391,7 +391,7 @@ class Comment(models.Model):
         self.text = text
         self.text_html = html
         self.save()
-        all_the_pings = list(metadata.get('ping', []))
+        all_the_pings = list(filter(lambda user_name: user_name != self.author.username, metadata.get('ping', [])))
         max_ping_count = settings.ZDS_APP['comment']['max_pings']
         first_pings = all_the_pings[:max_ping_count]
         for username in first_pings:

--- a/zds/utils/models.py
+++ b/zds/utils/models.py
@@ -386,18 +386,22 @@ class Comment(models.Model):
     hat = models.ForeignKey(Hat, verbose_name='Casquette', on_delete=models.SET_NULL,
                             related_name='comments', blank=True, null=True)
 
-    def update_content(self, text):
+    def update_content(self, text, on_error=None):
         _, old_metadata, _ = render_markdown(self.text)
-        html, metadata, messages = render_markdown(text)
+        html, metadata, messages = render_markdown(text, on_error=on_error)
         self.text = text
         self.text_html = html
         self.save()
         all_the_pings = list(filter(lambda user_name: user_name != self.author.username, metadata.get('ping', [])))
-        all_the_pings = list(set(all_the_pings) - set(old_metadata))
+        all_the_pings = list(set(all_the_pings) - set(old_metadata.get('ping', [])))
         max_ping_count = settings.ZDS_APP['comment']['max_pings']
         first_pings = all_the_pings[:max_ping_count]
         for username in first_pings:
             signals.new_content.send(sender=self.__class__, instance=self, user=User.objects.get(username=username))
+        unpinged_usernames = set(old_metadata.get('ping', [])) - set(all_the_pings)
+        unpinged_users = User.objects.filter(username__in=list(unpinged_usernames))
+        for unpinged_user in unpinged_users:
+            signals.unsubscribe.send(self.author, instance=self, user=unpinged_user)
 
     def hide_comment_by_user(self, user, text_hidden):
         """Hide a comment and save it

--- a/zds/utils/models.py
+++ b/zds/utils/models.py
@@ -386,12 +386,14 @@ class Comment(models.Model):
     hat = models.ForeignKey(Hat, verbose_name='Casquette', on_delete=models.SET_NULL,
                             related_name='comments', blank=True, null=True)
 
-    def update_content(self, text, on_error=None):
-        html, metadata, messages = render_markdown(text, on_error=on_error)
+    def update_content(self, text):
+        _, old_metadata, _ = render_markdown(self.text)
+        html, metadata, messages = render_markdown(text)
         self.text = text
         self.text_html = html
         self.save()
         all_the_pings = list(filter(lambda user_name: user_name != self.author.username, metadata.get('ping', [])))
+        all_the_pings = list(set(all_the_pings) - set(old_metadata))
         max_ping_count = settings.ZDS_APP['comment']['max_pings']
         first_pings = all_the_pings[:max_ping_count]
         for username in first_pings:

